Code/GraphMol/Descriptors/AtomicEnvironmentVector.cpp

```diff
@@ -0,0 +1,845 @@
+  \param dim        dimension along which indices are to be picked
+
+  \return Matrix containing values at positions specified by index in vector2
+*/
+template <typename Derived>
+void IndexSelect(ArrayBase<Derived> *vector1, ArrayBase<Derived> *vector2,
+                 ArrayXi &index, unsigned int dim) {
+  PRECONDITION(vector1 != nullptr && vector2 != nullptr,
+               "Input vectors are NULL");
+  PRECONDITION(dim == 0 || dim == 1,
+               "Only values 0 and 1 are accepted for dim");
+  for (auto i = 0; i < index.size(); i++) {
+    switch (dim) {
+      case 0:
+        vector1->row(i) = vector2->row(index(i));
+        break;
+      case 1:
+        vector1->col(i) = vector2->col(index(i));
+        break;
+      default:
+        throw ValueErrorException("Value of dim must be 0 or 1");
+    }
+  }
+}
+
+//! Computes pairs of atoms that are neighbors bypassing duplication to make
+//! calculation faster
+/*!
+  \param coordinates  A matrix of size atoms * 3 containing coordinates of each
+  atom
+  \param species      A vector of size atoms containing mapping from atom
+  index to encoding
+  \param cutoff       Maximum distance within which 2 atoms
+  are considered to be neighbours
+  \param atomIndex12  Array in which each column represents pairs of atoms
+
+  \return 2 dimensional array with 2 rows with each column corresponding to a
+  pair of atoms which are neighbours
+*/
+void NeighborPairs(ArrayXXd *coordinates, const VectorXi *species,
+                   double cutoff, unsigned int numAtoms, ArrayXi *atomIndex12) {
+  PRECONDITION(coordinates != nullptr, "Coordinates are NULL");
+  PRECONDITION(species != nullptr, "species vector is NULL");
+  PRECONDITION(coordinates->rows() == numAtoms,
+               "Number of coordinate vectors must be same as number of atoms");
+  PRECONDITION(species->size() == numAtoms,
+               "Size of species vector must be same as number of atoms");
+
+  // Find atoms which are not H, C, N or O
+  auto paddingMask = species->array() == -1;
+  unsigned int cols = 0;
+  auto numPairs = numAtoms * (numAtoms - 1) / 2;
+
+  // Each column contains pair of indices of the upper
+  // traingular matrix of size numAtoms
+  MatrixXi upperTriag(2, numPairs);
+  for (unsigned int i = 0; i < numAtoms; i++) {
+    for (unsigned int j = i + 1; j < numAtoms; j++) {
+      upperTriag.col(cols) << i, j;
+      cols++;
+    }
+  }
+
+  // Flattened previously mentioned matrix with indices
+  // Example :  (0, 0, 1) -> (0, 0, 1, 1, 2, 2)
+  //            (1, 2, 2)
+  ArrayXi upperTriagFlattened(numPairs * 2);
+  unsigned int index = 0;
+  for (auto i = 0; i < upperTriag.rows(); i++) {
+    for (auto j = 0; j < upperTriag.cols(); j++) {
+      upperTriagFlattened(index) = upperTriag(i, j);
+      index++;
+    }
+  }
+
+  // Select indices in upperTriagFlattened from coordinates
+  ArrayXXd pairCoordinates(numPairs * 2, 3);
+  IndexSelect(&pairCoordinates, coordinates, upperTriagFlattened, 0);
```

Comment:
If IndexSelect used references IndexSelect(ArrayBase<Derived> &vector1, ...) this would be safer code and would remove the nullptr PRECONDITIONs.